Added  "searx"  engine support

masterofobzene opened the pull request.

rany2 commented:
I guess this a step forward, but it would be nice for this to work on all Searx instances and allow the user to list them. Extensions have an ability to request permissions to access new sites so this could be used here.

Also revert your changes to the ReadMe

masterofobzene commented:
reverted readme.

Can you provide an example on how to do that?  I can't find any examples on the already working engines, only the match pattern is there.

masterofobzene commented:
Ok, I've added all public domains "by hand" since there is no "future proof" matching pattern possible. 
Also added some domains for vainilla matching patterns.
All the working domains are the ones listed in: https://searx.space/

iorate commented:
At least you should resolve the syntax error.

iorate commented:
@masterofobzene See my comment and [readme](https://github.com/iorate/ublacklist#build). Your code does not even pass the build.

masterofobzene commented:
> @masterofobzene See my comment and [readme](https://github.com/iorate/ublacklist#build). Your code does not even pass the build. Have you checked that your code works well???

Yes and it passes. 
![imagen](https://user-images.githubusercontent.com/94469116/152349668-f24d455a-0af9-477f-a676-1527a8a57066.png)

Aslo that error of the comma was there even before I touched the code.
![imagen](https://user-images.githubusercontent.com/94469116/152349838-0c60f3af-017b-489b-82c8-be8b0f9f929f.png)

masterofobzene commented:
This is your search-engines.ts file, it shows a different error.

![imagen](https://user-images.githubusercontent.com/94469116/152350327-0bad5522-e802-4474-99be-ea503a1b67c9.png)

masterofobzene commented:
and the extension compiled is working as expected:
![imagen](https://user-images.githubusercontent.com/94469116/152350533-8f45476e-1ec3-4ab8-91e8-7881aee4934d.png)


![imagen](https://user-images.githubusercontent.com/94469116/152350650-6f5138fc-356b-4cd4-986b-2c0a5acb87a7.png)

iorate commented:
The current master passed the check. https://github.com/iorate/ublacklist/runs/4999544004?check_suite_focus=true
Obviously you introduced the errors.

<s>Your `src/common/search-engines.ts` have a syntax error. Insert `},` *before* the code you added.</s>Seems done

I cannot understand why "a different error" occurred. `src/common/search-engines.ts` should not include the HTML code (`<!DOCTYPE html> ...`), unless you add it.

I guess many warnings by prettier are due to autocrlf by Git for Windows.

masterofobzene commented:
I don't know but if I added something it should be logged. I didn't (on your files).

masterofobzene commented:
Man, your code is giving the same errors, I just cloned the vainilla repo and ran lint:
![imagen](https://user-images.githubusercontent.com/94469116/152360319-1380f70e-2c6a-4939-96ea-78f843a99f33.png)

iorate commented:
> I guess many warnings by prettier are due to autocrlf by Git for Windows.

iorate commented:
To fix the remaining error in the check, run `yarn fix`.

masterofobzene commented:
> > I guess many warnings by prettier are due to autocrlf by Git for Windows.

I guess that notepad++ is breaking my search-engines.ts file.

iorate commented:
Run `yarn fix`, commit, and push.

iorate commented:
(Maybe I should introduce lint-staged and husky later)

masterofobzene commented:
ok as the newbie I am, I was running "yarn lint fix" instead of "yarn fix"  sorry.

masterofobzene commented:
now i'm puzzled. Can you run it on your end? don't know whats failing now.
![imagen](https://user-images.githubusercontent.com/94469116/152371560-aa843f44-bfca-4fb8-9fb1-b151bdf8a4a6.png)

![imagen](https://user-images.githubusercontent.com/94469116/152371628-804a7820-3a3b-41d3-8785-23c6b27fa0a0.png)

iorate commented:
Run `yarn fix` again and commit update of `src/scripts/search-engines/searx.ts`.

masterofobzene commented:
Now I really don't know whats failing. Started from scratch again, I cloned and ran all the commands and i get no errors.
Maybe someone on linux should run these tests.

![imagen](https://user-images.githubusercontent.com/94469116/152378153-e4cd8de7-ba5a-419f-a547-59c4be16ef50.png)

iorate commented on src/common/search-engines.ts:
```diff
@@ -302,4 +309,52 @@ export const SEARCH_ENGINES: Readonly<Record<SearchEngineId, Readonly<SearchEngi
+    contentScripts: [
+      {
+        matches: [
+          'https://*.searx.be/*',
```
Make the match patterns more narrow. Is access for every subdomain and page on `searx.be` really needed? Wouldn't `https://searx.be/search` (POST) and `https://searx.be/search?*` (GET) be enough?

The same applies to the following match patterns.

masterofobzene replied:
What would be the difference or benefit from that? (I ask honestly)

masterofobzene replied:
done, I will only use the POST method since the GET one would defeat the purpose of using Searx (privacy)

iorate commented on src/scripts/search-engines/searx.ts:
```diff
@@ -0,0 +1,49 @@
+import { SEARCH_ENGINES } from '../../common/search-engines';
```
This file is only a copy of `ecosia.ts`. Does it match the styles of the Searx instances?

I indicated some possible improvements below.

masterofobzene replied:
Yes, they worked, no weird stuff seen in my tests on firefox.

iorate commented on src/scripts/search-engines/searx.ts:
```diff
@@ -0,0 +1,49 @@
+        target: '.result',
+        url: 'a',
+        title: 'a',
+        actionTarget: root => root.querySelector('.result__body') || root,
```
This line includes unnecessary code for Searx. Replace this line with `actionTarget: ''`.

masterofobzene replied:
done

iorate commented on src/scripts/search-engines/searx.ts:
```diff
@@ -0,0 +1,49 @@
+        actionStyle: {
+          display: 'block',
+          fontSize: '13px',
+          paddingTop: '5px',
```
The "Block this site" links are too far from the corresponding results. You can simply remove this line, or replace it with a negative margin (e.g. `marginTop: '-10px'`).

masterofobzene replied:
done

iorate reviewed (changes requested):
Please check my comments.

iorate commented:
You should also update the readme and the website.
https://github.com/iorate/ublacklist/blob/master/README.md#supported-search-engines
https://github.com/iorate/ublacklist/blob/master/website/docs/introduction.md#features
https://github.com/iorate/ublacklist/blob/master/website/docs/advanced-features.md#other-search-engines

masterofobzene commented:
sorry I think I opened more push requests for the changes you asked.

iorate commented:
Please add the changes of readme and website to this pull request if possible.

iorate commented on src/common/search-engines.ts:
```diff
@@ -302,4 +309,52 @@ export const SEARCH_ENGINES: Readonly<Record<SearchEngineId, Readonly<SearchEngi
+    contentScripts: [
+      {
+        matches: [
+          'https://*.searx.be/search',
```
1. Searx supports the GET method (you can enable it in the preferences page). An example of GET URL is: https://searx.be/search?q=example&category_general=on&time_range=&language=all . So `https://searx.be/search?*` is also necessary.

2. Would you show an example of `searx.be`'s subdomain? Unnecessary permission for subdomains may lead to potential security problems. For example, if an Searx instance `searx.example` provides login functionality on `login.searx.example`, the users will not want to grant access to passwords that they enter on `login.searx.example` to this extension.
ref: https://developer.chrome.com/docs/extensions/mv3/user_privacy/#required_permissions

masterofobzene replied:
1. Using GET is a privacy hole. The idea of using Searx is to invisibilize your searches and using "privacy 0" mainstream search engines (Google) with a privacy wall that is searx itself. However GET sends the requested search as a complete URL and your DNS would leak your info with this method. This can be added to the readme file if you want.

2. There was one and I cannot recall which one was, but it was probably one of the instances that doesn't use google so its not being used here. What logging are you talking about?

masterofobzene commented:
> Please add the changes of readme and website to this pull request if possible.

done that

> Please add the changes of readme and website to this pull request if possible.

All the info has been added to the readme files. Check the pull requests.

masterofobzene commented:
> Please add the changes of readme and website to this pull request if possible.

done

iorate commented:
@masterofobzene ANSWER MY QUESTION. 
Is permission for subdomains (e.g. `*://*.searx.be/search`) really needed???

masterofobzene commented:
> @masterofobzene ANSWER MY QUESTION. Is permission for subdomains (e.g. `*://*.searx.be/search`) really needed???

How does the extension operate without the permission?

masterofobzene commented:
Why is this forcing programming formats on a readme file?

iorate commented:
ANSWER MY QUESTION. Is permission for subdomains (e.g. *://*.searx.be/search) really needed???
What I want to say is that `*://searx.be/search` would be enough.

masterofobzene commented:
> ANSWER MY QUESTION. Is permission for subdomains (e.g. _://_.searx.be/search) really needed??? What I want to say is that `*://searx.be/search` would be enough.

And how do you know if someone will just want to add a domain called someting.searx.be/search in the future?
I also added a lot of "vainilla" domains that do not exist right now just in case someone decides to use them.
Do want to delete them?

iorate commented:
https://developer.chrome.com/docs/extensions/mv3/user_privacy/#required_permissions

> Extensions should not "future proof" access to user data by requesting permissions that they do not currently need, but may implement in the future.

masterofobzene commented:
> https://developer.chrome.com/docs/extensions/mv3/user_privacy/#required_permissions
> 
> > Extensions should not "future proof" access to user data by requesting permissions that they do not currently need, but may implement in the future.

Ok, I will delete them.
Now, you are not expecting me to run lint on a, readme right?

iorate commented:
Run `yarn fix` in the `website` directory.

iorate commented:
Why did you remove the blank line before the table in readme to break it?
Please don't make me angry any more...

masterofobzene commented:
Metete la extension en el culo amigo.

DRSDavidSoft commented:
I'm planning to do another PR in order to merge this into master (issue: https://github.com/iorate/ublacklist/issues/143) thanks for the implementation hopefully it'll get added this time, everyone
